### PR TITLE
Add proc macro lib to compute code-block associated constants via hashing

### PR DESCRIPTION
**Pull request by CPerezz**
We need a way to version our circuits. A first step into
that direction was basically the fact of introducing the CIRCUIT_ID
associated const in the Circuit trait. See: https://github.com/dusk-network/plonk/pull/454

This is the second step which hashes the contents of the Circuit trait
implementation and hashes them into this constant. On that way, every
time the circuit changes, the ID will do the same. And we will be able
to catch this changes and version the circuits properly.

Resolves: #240

### CODE REVIEW

File: circuit-id-generator/tests/main.rs
Comment (ZER0) [inline circuit-id-generator/tests/main.rs:93]: The tests should checks:

1. Different structure with same circuit (same `gadget` and `padded_circuit_size`) ended up with the same Hash
2. Different structure with same `gadget` but different `padded_circuit_size` ended up with different hash
(we could also test different `gadget` but I think just checking one factor is enough)

If we use the generic approach:
1. Same structure different versioning ended up with different hash 
2. Test on code block that are not Circuit impl, with custom `const` name works

**Review by ZER0 — CHANGES_REQUESTED** [review]
If we want to land this, it needs changes.

Plus:
1. Add the CHANGELOG.md
2. Be sure every commits are atomic to the workspace member, so: you can have a single commit for any changes you're doing inside a member, but you cannot mix files outside of it. So for example here you should have one commit with all the new stuff you added, and one commit for the workspace's Cargo toml.
3. The commit should be prefixed with the member's name (unless is a change in the workspace root, so no prefix needed). See: https://github.com/tokio-rs/tokio/commits/master. For example if I'm changing something in rusk-abi, the commit message will be something like: `rusk-abi: Change foo in bar`

File: gadget_hasher/Cargo.toml
Comment (CPerezz) [reply gadget_hasher/Cargo.toml]: Nop/ It's a leftover. Removing it.

File: circuit-id-generator/src/lib.rs
Comment (CPerezz) [reply circuit-id-generator/src/lib.rs:16]: Makes sense! Looks to me that even if now we go for the generic approach. We will need to end up on a more complex impl with `syn` usage and aiming to tie up everything well and make the macro linked to plonk only.

Comment (CPerezz) [reply circuit-id-generator/src/lib.rs:16]: But this can be done in the future of course.

File: macros/Makefile
Comment (ZER0) [inline macros/Makefile:13]: Nit: add a new line at the end of the file.

File: macros/code-hasher/src/lib.rs
Comment (ZER0) [inline macros/code-hasher/src/lib.rs:32]: Nit: You're using both arguments now, so you should remove the underscore as prefix (so it should be `attr` and `input` instead of `_attr` and `_input`.

File: macros/code-hasher/src/lib.rs
Comment (ZER0) [inline macros/code-hasher/src/lib.rs:36]: Since you're using the (code) version attribute, you can avoid to pass the crate's version.
I suggested the crate's version only in case of non-generic hasher (so tied to Plonk), when you don't have the (code) version attribute.

Since you implemented, this can be omitted. If, instead, you prefer to add it, please add as mentioned both `major` and `minor`, otherwise if I have `0.4.0` and `1.4.0` it will result in the same hash (and it shouldn't).

**Review by ZER0 — APPROVED** [review]
Looks very good! Approved if all tests are passing, but there are still few nits to address before the merge, see comments.

Also, be careful: not all commits have the prefix `code-hasher:`, see: https://github.com/dusk-network/rusk/pull/249/commits

You either squash the commit (I don't mind) or you "reword" the ones that didn't have it yet and you add it.
But be extra careful, each commit of files inside a member needs to have the prefix, and you cannot have in one commit changes belongs to different members. For example this one:

https://github.com/dusk-network/rusk/pull/249/commits/e6b0db68bdcd3a24264f76ba462bc7e14efbdb5f

Contains changes both to the workspace root and the code-hasher. This should be split in two.

File: macros/code-hasher/src/lib.rs
Comment (CPerezz) [reply macros/code-hasher/src/lib.rs]: As discussed this will be removed for now. We can always add it later!

**Pull request merged**
